File: 2022/julia/day3.jl
function prios(items)
    p = merge(
        Dict(zip('a':'z', 1:26)),
        Dict(zip('A':'Z', 27:52))
    )
    getindex.(Ref(p), items)
end

function split(sack)
    n = div(length(sack), 2)
    sack[begin:n], sack[n+1:end]
end

function overlaps(sack)
    collect(intersect(Set.(collect.(sack))...))
end

function groups(lines)
    ixs = collect(3:3:length(lines)) .+ 1
    map(i -> reduce(vcat, lines[i-3:i-1]), ixs)
end

function sol1(lines)
    sum(prios.(reduce(vcat, overlaps.(split.(lines)))))
end

function sol2(lines)
    sum(prios.(reduce(vcat, overlaps.(groups(lines)))))
end


readlines("../input/3.txt") |> sol1 |> println
readlines("../input/3.txt") |> sol2 |> println

File: 2022/haskell/src/Day3.hs
module Day3 where

import Data.Maybe
import Data.List.Split
import qualified Data.Text as T
import qualified Data.Set as Set
import qualified Data.Map as Map


compartments xs = [p1, p2]
  where
    (p1, p2) = T.splitAt halfway xs
    halfway = T.length xs `div` 2

overlap = head . Set.elems . foldr1 Set.intersection . fmap (Set.fromList . T.unpack)

score x =
  fromJust $ Map.lookup x scoreMap
  where
    scoreMap = lowerCaseMap <> upperCaseMap
    lowerCaseMap = Map.fromList $ zip ['a'..'z'] [1..26]
    upperCaseMap = Map.fromList $ zip ['A'..'Z'] [27..52]

groups = chunksOf 3

sol1 = sum . fmap (score . overlap . compartments) . T.lines
sol2 = sum . fmap (score . overlap) . groups . T.lines
